ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol

```diff
@@ -1,623 +1,560 @@
-        // Amount of receiverValue that is refunded to the user (0 if the call to 'receiveWormholeMessages' did not revert, or the full receiverValue otherwise)
-        uint256 receiverValueRefundAmount =
-            (receiverValueWasPaid ? 0 : deliveryInstruction.receiverValueTarget);
-
-        // Total refund to the user
-        uint256 refundToRefundAddress =
-            receiverValueRefundAmount + transactionFeeRefundAmountPostForward;
-
-        // Whether or not the refund succeeded
-        bool refundPaidToRefundAddress;
-        (refundPaidToRefundAddress, refundStatus) =
-            payRefundToRefundAddress(deliveryInstruction, refundToRefundAddress, providerAddress);
-
-        // Refund the relayer (their extra funds) + (the amount that the relayer spent on gas)
-        // + (the users refund if that refund didn't succeed)
-        uint256 relayerRefundAmount = (
-            msg.value - deliveryInstruction.receiverValueTarget
-                - deliveryInstruction.maximumRefundTarget
-        ) + (deliveryInstruction.maximumRefundTarget - transactionFeeRefundAmount)
-            + (refundPaidToRefundAddress ? 0 : refundToRefundAddress);
-
-        Utils.pay(relayerRefundAddress, relayerRefundAmount);
+    //TODO AMO: should never revert for any other reason (though it might for overflows atm)
+    catch (bytes memory revertData) {
+      //decode returned Cancelled error
+      uint256 available;
+      uint256 required;
+      (gasUsed, available, required) = decodeCancelled(revertData);
+      //Calculate the amount of maxTransactionFee to refund (multiply the maximum refund by the
+      //  fraction of gas unused)
+      transactionFeeRefundAmount =
+        calculateTransactionFeeRefundAmount(vaaInfo.deliveryInstruction, gasUsed);
+      status = DeliveryStatus.FORWARD_REQUEST_FAILURE;
+      additionalStatusInfo =
+        abi.encodeWithSelector(ForwardNotSufficientlyFunded.selector, available, required);
+      transactionFeeRefundAmountPostForward = transactionFeeRefundAmount;
     }
 
-    function payRefundToRefundAddress(
-        IWormholeRelayerInternalStructs.DeliveryInstruction memory instruction,
-        uint256 refundAmount,
-        bytes32 relayerAddress
-    ) internal returns (bool refundPaidToRefundAddress, RefundStatus refundStatus) {
-        if (instruction.refundChain == chainId()) {
-            refundPaidToRefundAddress =
-                Utils.pay(payable(fromWormholeFormat(instruction.refundAddress)), refundAmount);
-            refundStatus =
-                refundPaidToRefundAddress ? RefundStatus.REFUND_SENT : RefundStatus.REFUND_FAIL;
-        } else {
-            address providerAddress = fromWormholeFormat(relayerAddress);
-            (refundPaidToRefundAddress, refundStatus) =
-                payRefundRemote(instruction, refundAmount, providerAddress);
-        }
-    }
+    //TODO AMO: At this point forwards (which are payable!) might have increased contract balance
```

Comment:
We store the msg value that the forwarder puts in, in state, and then later pay it out as part of the forward relayer's reward, so I don’t think we need to account for that here? Those funds are been payed out to the relay provider for the forward. Let me know if I am understanding this comment wrong.

Reply:
No, this is perfectly fine. I wrote this earlier on when I hadn't seen/digested the code in emitForwards yet, so this really is just outdated.